### PR TITLE
Dnnl refactor

### PR DESCRIPTION
**Pull request by georgen117**
**Description**:
This is a rework of the DNNL execution provider.

`DnnlTensor`, `DnnlNode`, and `DnnlSubgraph` have been updated to support arbitrary graph topology and tensor data types. The previous version of the DNNL execution provider only supported F32 data types.

rework `GetCapability` to claim nodes in graph greedily from node topological ordering and delay creation of `DnnlSubgraph` until Compile

rework compile to have `DnnlSubgraphPrimitive` as the object to handle primitive creation and execution instead of thread-local primitive pool which duplicated intermediate memory allocated by the execution provider across threads.

`DnnlSubgraphPrimitive` provides helpers to handle many common functions for each DNNL primitive builder and become the centralized place to:
- store input , output, intermediate memory, initializer memory, etc
- it provides functions to obtain input memories with automatic reordering/reshaping and moving between DNNL engines
- it provides interfaces to add primitive, set output memory for each node, etc
- Improved initializer caching

add `CONCURRENT_EXEC` compile flag for DNNL library as, without it, convolution primitive cannot be created and executed on different threads

Enable unit tests to run on DNNL execution provider as well if built with DNNL execution provider

Updated the following Operators for the reworked execution provider:
- AveragePool
- BatchNorm
- Conv
- ConvGrad
- GlobalAveragePool
- GlobalMaxPool
- LRN
- MatMul
- MaxPoolGrad
- MaxPool
- ReduceMean
- Relu
- ReluGrad
- Sum

Implemented new operators:
- Add
- AveragePoolGrad
- Div
- MatMulInteger
- Mul
- SoftMax
- Sub

**Motivation and Context**
- Why is this change required? What problem does it solve?

This PR solves several major issues:
- It make it possible to support more than just the Float32 data type.
- It improves DNNL subgraph generation:
    - It takes the topology of the graph into consideration more than the previous code
    - a subgraph can support more than one output.
- Memory (inputs, outputs, and reorders) are now handle in the `DnnlSubgraphPrimitive` removing the need to handle those in each operator.
    - This reduces need to worry about the DNNL engine (i.e. CPU vs. GPU) when writing each operator
    - Simplifies code size for each operator for example Relu went from 207 lines to 70, Conv went from 777 lines to 478.
- better code verification via unit test since the default unit tests now run the DNNL execution provider if the dnnl build flag is provided.

### CODE REVIEW

**Comment by jywu-msft** [comment]
seems like there's a conflict on dnnl_node_capability.cc ?

**Comment by georgen117** [comment]
> seems like there's a conflict on dnnl_node_capability.cc ?

I am fixing the conflict and will have it pushed up shortly.

**Comment by georgen117** [comment]
@jywu-msft thanks for running the test build.  We have 16 failures in the ModelTests.

I know from past experience that the models for those tests are not available publicly. We are not 100% sure but it appears that the failure may be in our Softmax implementation.  Which is one of the new Ops that was added that didn't exist in the previous DNNL EP.

I may need to make a new commit and ask you to run the Linux DNNL CI Pipeline with some extra debugging information turned on to get extra information needed to fix the failure.

**Comment by jywu-msft** [comment]
> @jywu-msft thanks for running the test build. We have 16 failures in the ModelTests.
> 
> I know from past experience that the models for those tests are not available publicly. We are not 100% sure but it appears that the failure may be in our Softmax implementation. Which is one of the new Ops that was added that didn't exist in the previous DNNL EP.
> 
> I may need to make a new commit and ask you to run the Linux DNNL CI Pipeline with some extra debugging information turned on to get extra information needed to fix the failure.

these are node tests from onnx which are public:
https://github.com/onnx/onnx/tree/master/onnx/backend/test/data/node

you can match the test names (and also the onnx version , looks like 1.8.1 and 1.7.0 )
1: [  FAILED  ] 16 tests, listed below:
1: [  FAILED  ] ModelTests/ModelTest.Run/4768, where GetParam() = "dnnl_/data/onnx/onnx181/node/test_sce_NCd1d2d3d4d5_mean_weight_log_prob/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/4776, where GetParam() = "dnnl_/data/onnx/onnx181/node/test_sce_NCd1d2d3d4d5_mean_weight_log_prob_expanded/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/4825, where GetParam() = "dnnl_/data/onnx/onnx181/node/test_nllloss_NCd1d2d3d4d5_mean_weight/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/4882, where GetParam() = "dnnl_/data/onnx/onnx181/node/test_sce_NCd1d2d3d4d5_none_no_weight_log_prob_expanded/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/4981, where GetParam() = "dnnl_/data/onnx/onnx181/node/test_sce_NCd1d2d3d4d5_mean_weight/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/5026, where GetParam() = "dnnl_/data/onnx/onnx181/node/test_sce_NCd1d2d3d4d5_none_no_weight_expanded/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/5048, where GetParam() = "dnnl_/data/onnx/onnx181/node/test_nllloss_NCd1d2d3d4d5_mean_weight_expanded/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/5128, where GetParam() = "dnnl_/data/onnx/onnx181/node/test_sce_NCd1d2d3d4d5_mean_weight_expanded/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/6069, where GetParam() = "dnnl_/data/onnx/onnx170/node/test_softmax_cross_entropy_input_shape_is_NCd1d2d3d4d5_mean_weight/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/6074, where GetParam() = "dnnl_/data/onnx/onnx170/node/test_softmax_cross_entropy_input_shape_is_NCd1d2d3d4d5_mean_weight_log_prob/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/6152, where GetParam() = "dnnl_/data/onnx/onnx170/node/test_softmax_cross_entropy_input_shape_is_NCd1d2d3d4d5_none_no_weight_expanded/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/6256, where GetParam() = "dnnl_/data/onnx/onnx170/node/test_softmax_cross_entropy_input_shape_is_NCd1d2d3d4d5_mean_weight_log_prob_expanded/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/6261, where GetParam() = "dnnl_/data/onnx/onnx170/node/test_negative_log_likelihood_loss_input_shape_is_NCd1d2d3d4d5_mean_weight/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/6314, where GetParam() = "dnnl_/data/onnx/onnx170/node/test_softmax_cross_entropy_input_shape_is_NCd1d2d3d4d5_mean_weight_expanded/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/6369, where GetParam() = "dnnl_/data/onnx/onnx170/node/test_negative_log_likelihood_loss_input_shape_is_NCd1d2d3d4d5_mean_weight_expanded/model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/6534, where GetParam() = "dnnl_/data/onnx/onnx170/node/test_softmax_cross_entropy_input_shape_is_NCd1d2d3d4d5_none_no_weight_log_prob_expanded/model.onnx"

**Comment by georgen117** [comment]
> these are node tests from onnx which are public:
https://github.com/onnx/onnx/tree/master/onnx/backend/test/data/node

Thank you @jywu-msft

**Comment by georgen117** [comment]
@jywu-msft with your pointer to the model we were quickly able to find the issue.

Would you please run the  `Linux DNNL CI Pipeline` test again?

**Comment by georgen117** [comment]
> /azp run Linux DNNL CI Pipeline

Thank you @jywu-msft it looks like we have some new test failures to investigate.

**Comment by georgen117** [comment]
@jywu-msft Would you please run the Linux DNNL CI Pipeline test again?

once again thanks for your help.

**Comment by georgen117** [comment]
@jywu-msft the Linux DNNL CI pipline build has passed.  What are our next steps?

I am asking this because we know this is a large code change that is not trivial to review.

**Comment by jywu-msft** [comment]
> @jywu-msft the Linux DNNL CI pipline build has passed. What are our next steps?
> 
> I am asking this because we know this is a large code change that is not trivial to review.

will kick off rest of CI's and look over the code.

**Comment by georgen117** [comment]
@jywu-msft the Windows GPU CI Pipeline is failing 
```
The following tests FAILED:
	  3 - winml_test_image (Failed)
```
I am not sure if we can even address this failure.  Unless something has gone really wrong the build does not even contain any of the dnnl excution provider code.

Do we need to look into this failure.  I don't think it is due to any changes by us.

**Comment by georgen117** [comment]
@jywu-msft is there anything we can do from our end to help with the failures.

Since it is in the dml instead of dnnl I don't think there is much we will be able to do to resolve the issues.

I needed I can rebase the code to the latest head.

**Comment by jywu-msft** [comment]
> @jywu-msft is there anything we can do from our end to help with the failures.
> 
> Since it is in the dml instead of dnnl I don't think there is much we will be able to do to resolve the issues.
> 
> I needed I can rebase the code to the latest head.

@snnn , are you aware of any issues with DML tests failing?

**Comment by georgen117** [comment]
@jywu-msft it looks like we are just waiting on a reviewer. let me know if there is anything I can do to help.

**Comment by jywu-msft** [comment]
> @jywu-msft it looks like we are just waiting on a reviewer. let me know if there is anything I can do to help.

we should be able to merge this in later today.

**Review by jywu-msft — APPROVED** [review]
